File: AccuPay/Domain/Payroll/SssCalculator.vb
Option Strict On

Imports AccuPay.Data
Imports AccuPay.Data.Enums
Imports AccuPay.Data.Services
Imports AccuPay.Entity

Public Class SssCalculator

    Private ReadOnly _settings As ListOfValueCollection

    Private ReadOnly _socialSecurityBrackets As IReadOnlyCollection(Of Entities.SocialSecurityBracket)

    Public Sub New(settings As ListOfValueCollection, socialSecurityBrackets As IReadOnlyCollection(Of Entities.SocialSecurityBracket))
        _settings = settings
        _socialSecurityBrackets = socialSecurityBrackets
    End Sub

    Public Sub Calculate(paystub As Paystub, previousPaystub As Entities.Paystub, salary As Entities.Salary, employee As Entities.Employee, payperiod As Entities.PayPeriod)
        ' Reset SSS values to zero
        paystub.SssEmployeeShare = 0
        paystub.SssEmployerShare = 0

        ' If salary is is set not to pay sss, return.
        If Not salary.DoPaySSSContribution Then
            Return
        End If

        ' Get the social security bracket based on the amount earned.
        Dim amount = GetSocialSecurityAmount(paystub, previousPaystub, salary, employee)
        Dim socialSecurityBracket = FindMatchingBracket(amount)

        ' If no bracket was matched/found, then there's nothing to compute.
        If socialSecurityBracket Is Nothing Then
            Return
        End If

        Dim employeeShare = socialSecurityBracket.EmployeeContributionAmount
        Dim employerShare = socialSecurityBracket.EmployerContributionAmount + socialSecurityBracket.EmployeeECAmount

        If employee.IsWeeklyPaid Then
            Dim shouldDeduct = If(employee.IsUnderAgency, payperiod.SSSWeeklyAgentContribSched, payperiod.SSSWeeklyContribSched)

            If shouldDeduct Then
                paystub.SssEmployeeShare = employeeShare
                paystub.SssEmployerShare = employerShare
            End If
        Else
            Dim deductionSchedule = employee.SssSchedule

            If IsSssPaidOnFirstHalf(payperiod, deductionSchedule) Or IsSssPaidOnEndOfTheMonth(payperiod, deductionSchedule) Then
                paystub.SssEmployeeShare = employeeShare
                paystub.SssEmployerShare = employerShare
            ElseIf IsSssPaidPerPayPeriod(deductionSchedule) Then
                paystub.SssEmployeeShare = employeeShare / CalendarConstants.SemiMonthlyPayPeriodsPerMonth
                paystub.SssEmployerShare = employerShare / CalendarConstants.SemiMonthlyPayPeriodsPerMonth
            End If
        End If
    End Sub

    Private Function FindMatchingBracket(amount As Decimal) As Entities.SocialSecurityBracket
        Return _socialSecurityBrackets.FirstOrDefault(Function(s) s.RangeFromAmount <= amount And s.RangeToAmount >= amount)
    End Function

    Private Function GetSocialSecurityAmount(paystub As Paystub,
                                             previousPaystub As Entities.Paystub,
                                             salary As Entities.Salary,
                                             employee As Entities.Employee) As Decimal

        Dim policyByOrganization = _settings.GetBoolean("Policy.ByOrganization", False)

        Dim calculationBasis = _settings.GetEnum("SocialSecuritySystem.CalculationBasis", SssCalculationBasis.BasicSalary, policyByOrganization)

        Select Case calculationBasis
            Case SssCalculationBasis.BasicSalary

                Return Data.Helpers.PayrollTools.GetEmployeeMonthlyRate(employee, salary)

            Case SssCalculationBasis.Earnings

                Return If(previousPaystub?.TotalEarnings, 0) + paystub.TotalEarnings

            Case SssCalculationBasis.GrossPay

                Return If(previousPaystub?.GrossPay, 0) + paystub.GrossPay

            Case SssCalculationBasis.BasicMinusDeductions

                Return If(previousPaystub?.TotalDaysPayWithOutOvertimeAndLeave, 0) + paystub.TotalDaysPayWithOutOvertimeAndLeave

            Case SssCalculationBasis.BasicMinusDeductionsWithoutPremium

                Dim totalHours = If(previousPaystub?.TotalWorkedHoursWithoutOvertimeAndLeave, 0) + paystub.TotalWorkedHoursWithoutOvertimeAndLeave

                If (New SystemOwnerService()).GetCurrentSystemOwner() = SystemOwnerService.Benchmark AndAlso employee.IsPremiumInclusive Then

                    totalHours = If(previousPaystub?.RegularHoursAndTotalRestDay, 0) + paystub.RegularHoursAndTotalRestDay
                End If

                Dim monthlyRate = Data.Helpers.PayrollTools.GetEmployeeMonthlyRate(employee, salary)
                Dim dailyRate = Data.Helpers.PayrollTools.GetDailyRate(monthlyRate, employee.WorkDaysPerYear)
                Dim hourlyRate = Data.Helpers.PayrollTools.GetHourlyRateByDailyRate(dailyRate)

                Return totalHours * hourlyRate

            Case Else

                Return 0
        End Select
    End Function

    Private Function IsSssPaidOnFirstHalf(payperiod As Entities.PayPeriod, deductionSchedule As String) As Boolean
        Return payperiod.IsFirstHalf And (deductionSchedule = ContributionSchedule.FIRST_HALF)
    End Function

    Private Function IsSssPaidOnEndOfTheMonth(payperiod As Entities.PayPeriod, deductionSchedule As String) As Boolean
        Return payperiod.IsEndOfTheMonth And (deductionSchedule = ContributionSchedule.END_OF_THE_MONTH)
    End Function

    Private Function IsSssPaidPerPayPeriod(deductionSchedule As String) As Boolean
        Return deductionSchedule = ContributionSchedule.PER_PAY_PERIOD
    End Function

    Private Class ContributionSchedule
        Public Const FIRST_HALF As String = "First half"
        Public Const END_OF_THE_MONTH As String = "End of the month"
        Public Const PER_PAY_PERIOD As String = "Per pay period"
    End Class

End Class

File: Accupay.Data/Services/PayrollGeneration/SssCalculator.cs
using AccuPay.Data.Entities;
using AccuPay.Data.Enums;
using AccuPay.Data.Helpers;
using System.Collections.Generic;
using System.Linq;

namespace AccuPay.Data.Services
{
    public class SssCalculator
    {
        private readonly ListOfValueCollection _settings;

        private readonly IReadOnlyCollection<SocialSecurityBracket> _socialSecurityBrackets;

        public SssCalculator(ListOfValueCollection settings, IReadOnlyCollection<SocialSecurityBracket> socialSecurityBrackets)
        {
            _settings = settings;
            _socialSecurityBrackets = socialSecurityBrackets;
        }

        public void Calculate(Paystub paystub,
                            Paystub previousPaystub,
                            Salary salary,
                            Employee employee,
                            PayPeriod payperiod,
                            string currentSystemOwner)
        {
            // Reset SSS values to zero
            paystub.SssEmployeeShare = 0;
            paystub.SssEmployerShare = 0;

            // If salary is is set not to pay sss, return.
            if (!salary.DoPaySSSContribution)
                return;

            // Get the social security bracket based on the amount earned.
            var amount = GetSocialSecurityAmount(paystub,
                                                previousPaystub,
                                                salary,
                                                employee,
                                                currentSystemOwner);
            var socialSecurityBracket = FindMatchingBracket(amount);

            // If no bracket was matched/found, then there's nothing to compute.
            if (socialSecurityBracket == null)
                return;

            var employeeShare = socialSecurityBracket.EmployeeContributionAmount;
            var employerShare = socialSecurityBracket.EmployerContributionAmount + socialSecurityBracket.EmployeeECAmount;

            if (employee.IsWeeklyPaid)
            {
                var shouldDeduct = employee.IsUnderAgency ? payperiod.SSSWeeklyAgentContribSched : payperiod.SSSWeeklyContribSched;

                if (shouldDeduct)
                {
                    paystub.SssEmployeeShare = employeeShare;
                    paystub.SssEmployerShare = employerShare;
                }
            }
            else
            {
                var deductionSchedule = employee.SssSchedule;

                if (IsSssPaidOnFirstHalf(payperiod, deductionSchedule) ||
                    IsSssPaidOnEndOfTheMonth(payperiod, deductionSchedule))
                {
                    paystub.SssEmployeeShare = employeeShare;
                    paystub.SssEmployerShare = employerShare;
                }
                else if (IsSssPaidPerPayPeriod(deductionSchedule))
                {
                    paystub.SssEmployeeShare = employeeShare /
                                                CalendarConstants.SemiMonthlyPayPeriodsPerMonth;
                    paystub.SssEmployerShare = employerShare /
                                                CalendarConstants.SemiMonthlyPayPeriodsPerMonth;
                }
            }
        }

        private SocialSecurityBracket FindMatchingBracket(decimal amount)
        {
            return _socialSecurityBrackets.FirstOrDefault(s => s.RangeFromAmount <= amount &&
                                                                s.RangeToAmount >= amount);
        }

        private decimal GetSocialSecurityAmount(Paystub paystub,
                                                Paystub previousPaystub,
                                                Salary salary,
                                                Employee employee,
                                                string currentSystemOwner)
        {
            var policyByOrganization = _settings.GetBoolean("Policy.ByOrganization", false);

            var calculationBasis = _settings.GetEnum("SocialSecuritySystem.CalculationBasis", SssCalculationBasis.BasicSalary, policyByOrganization);

            switch (calculationBasis)
            {
                case SssCalculationBasis.BasicSalary:
                    return PayrollTools.GetEmployeeMonthlyRate(employee, salary);

                case SssCalculationBasis.Earnings:
                    return (previousPaystub?.TotalEarnings ?? 0) + paystub.TotalEarnings;

                case SssCalculationBasis.GrossPay:
                    return (previousPaystub?.GrossPay ?? 0) + paystub.GrossPay;

                case SssCalculationBasis.BasicMinusDeductions:
                    return (previousPaystub?.TotalDaysPayWithOutOvertimeAndLeave ?? 0) +
                            paystub.TotalDaysPayWithOutOvertimeAndLeave;

                case SssCalculationBasis.BasicMinusDeductionsWithoutPremium:
                    var totalHours = (previousPaystub?.TotalWorkedHoursWithoutOvertimeAndLeave ?? 0) +
                                        paystub.TotalWorkedHoursWithoutOvertimeAndLeave;

                    if (currentSystemOwner == SystemOwnerService.Benchmark && employee.IsPremiumInclusive)
                    {
                        totalHours = (previousPaystub?.RegularHoursAndTotalRestDay ?? 0) +
                                        paystub.RegularHoursAndTotalRestDay;
                    }

                    var monthlyRate = PayrollTools.GetEmployeeMonthlyRate(employee, salary);
                    var dailyRate = PayrollTools.GetDailyRate(monthlyRate, employee.WorkDaysPerYear);
                    var hourlyRate = PayrollTools.GetHourlyRateByDailyRate(dailyRate);

                    return totalHours * hourlyRate;

                default:
                    return 0;
            }
        }

        private bool IsSssPaidOnFirstHalf(PayPeriod payperiod, string deductionSchedule)
        {
            return payperiod.IsFirstHalf && deductionSchedule == ContributionSchedule.FIRST_HALF;
        }

        private bool IsSssPaidOnEndOfTheMonth(PayPeriod payperiod, string deductionSchedule)
        {
            return payperiod.IsEndOfTheMonth && deductionSchedule == ContributionSchedule.END_OF_THE_MONTH;
        }

        private bool IsSssPaidPerPayPeriod(string deductionSchedule)
        {
            return deductionSchedule == ContributionSchedule.PER_PAY_PERIOD;
        }
    }
}